### DISCUSSION TITLE
How to make the modified UI interface take effect on port 8080？？？

### DISCUSSION THREAD

**Discussion by c2006878** [nocodb/nocodb#6860] [Q&A]
I have made arbitrary changes to the UI and found that it can take effect on port 3000, but it does not take effect on port 8080, resulting in the inability to obtain the latest page when building the image. What configuration changes are needed to make it effective on port 8080?

**Comment by c2006878** [comment]
I am using a Windows system

**Comment by wingkwong** [comment]
Our backend (the one running on 8080 locally) takes the frontend part from the package `nc-lib-gui` which is the built version of `nc-gui` (the one running on 3000 locally). Therefore, every local changes you make on `nc-gui` can be only seen in port 3000.

**Reply by c2006878** [reply]
But how can I see the latest UI page by building a docker image locally?

**Reply by c2006878** [reply]
I am currently building a local docker image through the. sh file and accessing port 8080 without seeing any changes to the page. Is there any solution?

**Reply by wingkwong** [reply]
List I said before, the backend takes the built version of the frontend. In docker, we've built the frontend there so the changes should be reflected in docker. 

> accessing port 8080 without seeing any changes to the page. Is there any solution?

If you are building local docker through `build-local-docker-image.sh`, you should access port 3333 instead of 8080. I doubt 8080 is your local backend. You may double check this part.

**Reply by wingkwong** [reply]
Just have a quick test. It works and I could see the frontend change on local docker.

![image](https://github.com/nocodb/nocodb/assets/35857179/dbab037a-2f4c-4f8c-9ea1-08db5114f3f3)

**Reply by c2006878** [reply]
Why do we need to have two identical pages (front-end and back-end)? Should we access 3333 or 8080 in the final deployment testing?

**Reply by wingkwong** [reply]
There are no two identical pages. The page is from the built frontend package. In docker, we combine both frontend and backend as a single package and expose port 8080 internally. In `build-local-docker-image.sh,`, we show the message that you can run the following command. 

```
docker run -d -p 3333:8080 --name nocodb-local nocodb-local
```

In fact, this is just an example. You can still use 8080 if you want. Just change the command to 

```
docker run -d -p 8080:8080 --name nocodb-local nocodb-local
```

**Reply by c2006878** [reply]
Can you tell me the specific steps or instructions for "build front-end package"?
I just tested and there are changes to the local 3000 port. However, there was no change in accessing port 3333 after the image was built. I'm going crazy.
![1698907114787](https://github.com/nocodb/nocodb/assets/58558960/0008daa3-2b0b-4cf5-aef4-624c2e586621)
![1698907150250](https://github.com/nocodb/nocodb/assets/58558960/1606ca9a-5cda-473a-ac06-d40a46397fc3)

**Reply by da-da-da** [reply]
Could you please tell me how the backend can get the built frontend package? I assume the built frontend package is stored somewhere in the .packages/nocodb folder during the webpack stage. Also, could you please inform me if there's a command in the nc-gui/package.json file that I can use to build a frontend package for the backend?

**Reply by wingkwong** [reply]
@c2006878 It is covered in `build-local-docker-image.sh`. What changes did you make?

**Reply by wingkwong** [reply]
@da-da-da the backend takes `nc-lib-gui` package which is the built version of the frontend (`nc-gui`).

**Reply by c2006878** [reply]
I just changed the pages in the nc-gui folder (such as packages/nc-gui/components/dashboard/Sidebar/UserInfo. vue). Then building the image is execution ./build-local-docker-image.sh. Then access port 3333. But it just doesn't work.
![1698913287464](https://github.com/nocodb/nocodb/assets/58558960/c423f0e2-1b84-4193-8571-abcc7497b187)
![1698913356377](https://github.com/nocodb/nocodb/assets/58558960/d07082ce-39bc-428a-b47e-541d2bd91044)

**Reply by da-da-da** [reply]
Is `build:copy`  of `nc-gui` used to place built version to `nc-lib-gui` ?  
Plus, I also use `build-local-docker-image.sh`  to build the image, there are also nothing changed in only EXPOSE 8080 in docker. I wanna know how to modify UI in `nc-gui` and show the modified when I visit localhost:8080 in docker or just start backend locally.(Mabey I need know how to change `nc-lib-gui` when `nc-gui` is modified)

I think backend used old `nc-lib-gui`, not new built version.

**Reply by wingkwong** [reply]
@da-da-da 

> Is build:copy of nc-gui used to place built version to nc-lib-gui ?

No. This command simply build the frontend and copy to `nc-lib-gui/lib/dist/`. In production, we publish to npm. In local docker, we copy dist to `nocodb/docker/nc-gui` and webpack build. You can see the sh script. 

> Plus, I also use build-local-docker-image.sh to build the image, there are also nothing changed in only EXPOSE 8080 in docker. I wanna know how to modify UI in nc-gui and show the modified 

Basically I've already shared the approach. Just update `nc-gui` and run the script. I've tested today and shown the result above. 

> when I visit localhost:8080 in docker 

did you run `docker run -d -p 8080:8080 --name nocodb-local nocodb-local`? Please make sure 8080 is running by docker one instead of local backend.

> or just start backend locally.(Mabey I need know how to change nc-lib-gui when nc-gui is modified)

This won't work. If you use docker, you don't need to start your local backend.

**Reply by wingkwong** [reply]
@c2006878 Well I couldn't reproduce. Are you using the latest develop branch? Can you go to Terminal and run these commands to compare?

<img width="702" alt="image" src="https://github.com/nocodb/nocodb/assets/35857179/5a854794-4bb6-4a8e-8ccd-4fb836307e8c">

**Comment by da-da-da** [comment]
I have followed the steps in the link https://docs.nocodb.com/engineering/development-setup. I believe that if we are able to visit localhost:8080 with the new UI modification, it could solve our issue. Reason being, we are unable to get the new modification of `nc-gui` following the development-setup or building image in docker.

**Reply by wingkwong** [reply]
If it's running locally, then it's not possible as `packages/nocodb` takes the built version of frontend from npm unless you point it to the local build path.

**Comment by xutao0317** [comment]
Hello, how can we solve this problem locally? After modifying the code on nc gui, it can be displayed on port 3000. How can I synchronize the modifications on port 8080?

**Reply by xutao0317** [reply]
@wingkwong

**Reply by wingkwong** [reply]
Please read the above thread.

**Reply by xutao0317** [reply]
Thank you for your reply. I followed the operation of readme in nc lib gui,
# nc-lib-gui
This package is the build version of ``nc-gui`` which will be used in ``packages/nocodb``. The built files should be located in ``packages/nc-lib-gui/lib/dist/``. 
To build the frontend and copy the built files to dist directory, run ``npm run build:copy`` under ``packages/nc-gui/``.
But it still hasn't taken effect. I didn't do it in Docker, I just operated it in Windows

**Reply by xutao0317** [reply]
@wingkwong

**Reply by wingkwong** [reply]
https://github.com/nocodb/nocodb/discussions/6860#discussioncomment-7453878

**Comment by Aaron-Lutz** [comment]
has any one figured this out? We are still getting the same problem when following all the 'correct' steps. it seems to be copying the files over to the nc-gui dir, however it seems that the changes made to the source code dont stick or are not properly built with nuxt.
